### PR TITLE
Add support for Poisson2Sparse

### PR DESCRIPTION
**Pull request by jscanvic**
Poisson2Sparse is a self-supervised method for Poisson denoising combining non-linear sparse coding parametrized by a small neural network (à la Deep Image Prior) and Neighbor2Neighbor.

I suggest to add support for it in the library and this PR implements it.

### CODE REVIEW

**Comment by jscanvic** [comment]
The algorithm uses the proximal operator of a L1 loss weighted entry-wise - could be a use-case for #422

File: examples/self-supervised-learning/demo_poisson2sparse.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+r"""
+Poisson denoising using Poisson2Sparse
+=======================================
+
+This code shows how to restore an image corrupted by Poisson noise using Poisson2Sparse
```
Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py]:
```suggestion
This code shows how to restore a single image corrupted by Poisson noise using Poisson2Sparse, without requiring external training.
```

File: examples/self-supervised-learning/demo_poisson2sparse.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+"""
+
+import deepinv as dinv
+from deepinv.utils.plotting import plot
+import torch
+import deepinv as dinv
+import torch
+import deepinv as dinv
+import numpy as np
+import random
```
Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py]:
```suggestion
import deepinv as dinv
import torch
import numpy as np
import random
```

File: examples/self-supervised-learning/demo_poisson2sparse.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+# -------------
+
+x_hat = model(y)
+
+# Compute PSNR
+print(f"Measurement PSNR: {dinv.metric.PSNR()(x, y).item():.2f} dB")
+print(f"Poisson2Sparse PSNR: {dinv.metric.PSNR()(x, x_hat).item():.2f} dB")
+
+# Plot results
+plot([y, x_hat, x], titles=["Measurement", "Poisson2Sparse", "Ground truth"])
```
Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py]:
```suggestion
dinv.utils.plot([y, x_hat, x], titles=["Measurement", "Poisson2Sparse", "Ground truth"])
```

Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py]: add the PSNRs as subtitles of the plot, right now you can't see them

File: examples/self-supervised-learning/demo_poisson2sparse.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+).to(device)
+
+# %%
+# Run the model
+# -------------
+
+x_hat = model(y)
+
+# Compute PSNR
+print(f"Measurement PSNR: {dinv.metric.PSNR()(x, y).item():.2f} dB")
```
Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py]: swap around x and x_hat

File: examples/self-supervised-learning/demo_poisson2sparse.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    weight_n2n=2.0,
+    weight_l1_regularization=1e-5,
+    verbose=True,
+).to(device)
+
+# %%
+# Run the model
+# -------------
+
+x_hat = model(y)
```
Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py:65]: maybe explain that this doesn't require knowledge of physics/gain?

File: deepinv/models/poisson2sparse.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        self.weight_l1_regularization = weight_l1_regularization
+        self.num_iter = num_iter
+        self.verbose = verbose
+        self._loss_fn = self._Poisson2SparseLoss(
+            weight_n2n=self.weight_n2n,
+            weight_l1_regularization=self.weight_l1_regularization,
+        )
+
+    @_pad_fn_even
+    def forward(self, y, physics=None):
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: add kwargs?

File: examples/self-supervised-learning/demo_poisson2sparse.py (hunk 1 above)
Comment (Andrewwango) [reply examples/self-supervised-learning/demo_poisson2sparse.py]: without knowledge of the noise level

File: examples/self-supervised-learning/demo_poisson2sparse.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+# %%
+# Load a Poisson corrupted image
+# ------------------------------
+#
+# This example uses an image from the microscopy dataset FMD :footcite:t:`zhang2018poisson`.
+
+# Seed the RNGs for reproducibility
+torch.manual_seed(0)
+torch.cuda.manual_seed(0)
+random.seed(0)
```
Comment (Andrewwango) [inline examples/self-supervised-learning/demo_poisson2sparse.py]: I don't think we should need the import random,the random seed, the import numpy nor the the numpy seed. The implementation should be reproducible from torch seed alone. That being said, it would be even better to use torch generator and pass to the physics (and pass to the model if appropriate)

File: deepinv/tests/test_models.py [hunk 8]
```diff
@@ -775,24 +775,47 @@ def test_icnn(device, rng):
...
-#     )
-#     x = torch.ones(imsize, device=device).unsqueeze(0)
-#     y = physics(x)
-#     mse_in = (y - x).pow(2).mean()
-#     x_net = f(y, physics)
-#     mse_out = (x_net - x).pow(2).mean()
-#
-#     assert mse_out < mse_in
+@pytest.mark.parametrize("model_kind", ["DIP", "Poisson2Sparse"])
+def test_dip_like(model_kind, imsize, device):
```
Comment (Andrewwango) [inline deepinv/tests/test_models.py:863]: nice! long overdue!

File: deepinv/models/poisson2sparse.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
+# Adapted from https://github.com/tacalvin/Poisson2Sparse
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: include these attributions into the docstrings to be more visible + more precise

File: deepinv/models/poisson2sparse.py [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+# and https://github.com/drorsimon/CSCNet
+from __future__ import annotations
+from .base import Denoiser
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]:
```suggestion
from deepinv.models.base import Denoiser
```

File: deepinv/models/poisson2sparse.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+import torch.nn.functional as F
+from tqdm import trange
+from deepinv.loss import Neighbor2Neighbor as N2N
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: am a bit worried about future circular imports - maybe put this import in the `_Poisson2SparseLoss.__init__: import ; self.N2N = ...` and then in the forward `self.N2N....`

File: deepinv/models/poisson2sparse.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+from functools import wraps
+from typing import Callable
+from deepinv.utils.compat import zip_strict
+
+
+# Credit goes to https://github.com/drorsimon/CSCNet
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: again move the credit into the docstrings

File: deepinv/models/poisson2sparse.py [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+# Simon, Dror, and Michael Elad. "Rethinking the CSC model for natural images." Advances in Neural Information Processing Systems 32 (2019).
+
+
+class ConvLista(nn.Module):
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: docstring or make it a private class

File: deepinv/models/poisson2sparse.py [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+            x_pad = F.pad(y, (0, H % 2, 0, H % 2), value=value)
+            x_hat = fn(args[0], x_pad, *args[2:], **kwargs)
+            return x_hat[..., :H, :W]
+
+        return wrapper
+
+    return _decorate(func)
+
+
+class Poisson2Sparse(Denoiser):
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:255]: docstring + add to user guide + API

File: deepinv/models/poisson2sparse.py [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+    def __init__(
+        self,
+        *,
+        backbone: torch.nn.Module,
+        lr: float,
+        weight_n2n: float,
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: add default values

File: deepinv/models/poisson2sparse.py [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+
+    return _decorate(func)
+
+
+class Poisson2Sparse(Denoiser):
+    def __init__(
+        self,
+        *,
+        backbone: torch.nn.Module,
+        lr: float,
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: default suggested value

File: deepinv/models/poisson2sparse.py [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+        return wrapper
+
+    return _decorate(func)
+
+
+class Poisson2Sparse(Denoiser):
+    def __init__(
+        self,
+        *,
+        backbone: torch.nn.Module,
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: maybe add convlista as default backbone? + allow it to be passed as an arg

File: deepinv/models/poisson2sparse.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+        for _ in trange(self.num_iter, disable=not self.verbose):
+            loss, x_hat = self._loss_fn(y=y, model=backbone)
+
+            loss.backward()
+
+            optimizer.step()
+            optimizer.zero_grad()
+
+            with torch.no_grad():
+                x_hat = x_hat
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: what does thie achieve?

File: deepinv/models/poisson2sparse.py [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+                if x_hat_avg is None:
+                    x_hat_avg = x_hat
+                else:
+                    exp_weight = 0.98
+                    x_hat_avg = x_hat_avg * exp_weight + x_hat * (1 - exp_weight)
+                x_hat_avg = x_hat_avg.detach()
+
+        return x_hat_avg
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:323]:
```suggestion
        return x_hat_avg.detach()
```

File: deepinv/models/poisson2sparse.py [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+            y2_hat = torch.clamp(y2_hat, 0.0, 1.0)
+
+            # $\mathcal L_{\text{Poisson}}$
+            loss = self.pll_loss_fn(y2_hat, y2)
+
+            # $\mathcal L_{\mathrm{L1}}$
+            loss += F.l1_loss(y2_hat, y2)
+
+            # $\mathcal L_{\mathrm{N}}$
+            loss += self.weight_n2n * (((y2_hat - y2) - (x_hat1 - x_hat2)) ** 2).mean()
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:359]: use a norm method?

File: deepinv/models/poisson2sparse.py [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+            threshold = self.thresholds.view(1, -1, 1, 1)
+            return self._soft_threshold(x, threshold=threshold)
+
+        @staticmethod
+        def _soft_threshold(
+            x: torch.Tensor, *, threshold: float | torch.Tensor
+        ) -> torch.Tensor:
+            """Soft-thresholding operation
+
+            1. Beck, A., & Teboulle, M. (2009). A Fast Iterative Shrinkage-Thresholding Algorithm for Linear Inverse Problems. SIAM Journal on Imaging Sciences, 2(1), 183–202. https://doi.org/10.1137/080716542
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: proper citation

File: deepinv/models/poisson2sparse.py [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+        def forward(self, x):
+            # Channel-wise soft-thresholding with learned thresholds
+            threshold = self.thresholds.view(1, -1, 1, 1)
+            return self._soft_threshold(x, threshold=threshold)
+
+        @staticmethod
+        def _soft_threshold(
+            x: torch.Tensor, *, threshold: float | torch.Tensor
+        ) -> torch.Tensor:
+            """Soft-thresholding operation
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:231]: not sure such a a simple method requires a citation

File: deepinv/models/poisson2sparse.py [hunk 23]
```diff
@@ -0,0 +1,334 @@
...
+                    y.shape[1],
+                    y.shape[2] + top_pad + bot_pad,
+                    y.shape[3] + left_pad + right_pad,
+                ),
+                dtype=y.dtype,
+                device=y.device,
+            )
+            mask = torch.zeros_like(augmented_y)
+            for num, (augmented_y_part, mask_part) in enumerate(
+                zip_strict(augmented_y.unbind(dim=1), mask.unbind(dim=1))
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:153]: every day I learn a new torch method

File: deepinv/models/poisson2sparse.py [hunk 24]
```diff
@@ -0,0 +1,334 @@
...
+                    if (y.shape[2] + stride - kernel_size) % stride == 0
+                    else 2 * stride - ((y.shape[2] + stride - kernel_size) % stride)
+                ),
+            )
+            augmented_y = torch.empty(
+                (
+                    y.shape[0],
+                    stride**2,
+                    y.shape[1],
+                    y.shape[2] + top_pad + bot_pad,
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: shape[-2] and shape[-1] I think for generality

File: deepinv/models/poisson2sparse.py [hunk 25]
```diff
@@ -0,0 +1,334 @@
...
+# Simon, Dror, and Michael Elad. "Rethinking the CSC model for natural images." Advances in Neural Information Processing Systems 32 (2019).
+
+
+class ConvLista(nn.Module):
+    def __init__(
+        self,
+        *,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: int,
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: add default values

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Thanks for the contribution, it's cool to have more test time training methods! Some initial comments

**Comment by jscanvic** [comment]
Thanks for the review Andrew - will make the changes asap!

File: deepinv/models/poisson2sparse.py (hunk 19 above)
Comment (jscanvic) [reply deepinv/models/poisson2sparse.py:323]: It's already detached at this point so I assume you mean to get rid of line 289 as well? I can have a closer look at it if needed but iirc having the detach in the loop halves the memory footprint (at some point you have the old x_hat and the new x_hat in memory with their gradient & computational graph otherwise)

I made it a bit more compact by putting the detach in the 1st branch of the if/else (only place needed)

File: deepinv/models/poisson2sparse.py (hunk 20 above)
Comment (jscanvic) [reply deepinv/models/poisson2sparse.py:359]: Will have a look (there might be an error in the formula too)

Comment (jscanvic) [reply deepinv/models/poisson2sparse.py:359]: Okay the formula actually looks good there's just a typo in the paper (from: https://github.com/tacalvin/Poisson2Sparse/blob/f2b18acf755a204f44964cdbd8f1467cbcd8f991/src/inference.py#L127)

I don't think it makes sense to use a norm here, we'd have to weight the norm manually by 1/numel and undo the square root by squaring the norm and I don't think we gain much by doing that anyway?

File: deepinv/models/poisson2sparse.py [hunk 26]
```diff
@@ -0,0 +1,361 @@
...
+from functools import wraps
+from typing import Callable
+from deepinv.utils.compat import zip_strict
+
+
+class ConvLista(nn.Module):
+    r"""
+    Convolutional LISTA network.
+
+    The architecture was introduced by :footcite:t:`simon2019rethinking`, and it is wel suited as a backbone for Poisson2Sparse (see :class:`deepinv.models.Poisson2Sparse`).
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]:
```suggestion
    The architecture was introduced by :footcite:t:`simon2019rethinking`, and it is well suited as a backbone for Poisson2Sparse (see :class:`deepinv.models.Poisson2Sparse`).
```

File: deepinv/models/poisson2sparse.py [hunk 27]
```diff
@@ -0,0 +1,361 @@
+from __future__ import annotations
+from deepinv.models.base import Denoiser
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]: move to bottom

File: deepinv/models/poisson2sparse.py [hunk 28]
```diff
@@ -0,0 +1,361 @@
...
+        x = self.conv_C(gamma)
+
+        # Deal with the shifted copies
+        x = torch.masked_select(x, mask.bool())
+        x = x.reshape(shape[0], stride**2, *shape[1:])
+        # Average over all the shifts
+        x = x.mean(dim=1, keepdim=False)
+
+        # Post-processing
+        return torch.clamp(x, min=0.0, max=1.0)
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:119]: maybe we need a note saying that it requires tensors in range 0-1?

File: deepinv/models/poisson2sparse.py [hunk 29]
```diff
@@ -0,0 +1,361 @@
...
+    r"""
+    Poisson2Sparse model for Poisson denoising.
+
+    This method, introduced by :footcite:t:`ta2022poisson2sparse`, reconstructs an image corrupted by Poisson noise by learning a sparse non-linear dictionary parametrized by a neural network using a combination of Neighbor2Neighbor :footcite:t:`huang2021neighbor2neighbor`, of the negative log Poisson likelihood, of the :math:`\ell^1` pixel distance and of a sparsity-inducing :math:`\ell^1` regularization function on the weights.
+
+    .. note::
+
+        This method does not use knowledge of the physics model and assumes a Poisson degradation model internally. Therefore, the physics object can be omitted when calling the model and specifying it will have no effect.
+
+    :param torch.nn.Module | None backbone: Neural network used as a non-linear dictionary. If ``None``, a default :class:`deepinv.models.ConvLista` model is used.
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py]:
```suggestion
    :param torch.nn.Module, None backbone: Neural network used as a non-linear dictionary. If ``None``, a default :class:`deepinv.models.ConvLista` model is used.
```

File: deepinv/models/poisson2sparse.py [hunk 30]
```diff
@@ -0,0 +1,361 @@
...
+
+            # Stop gradients
+            with torch.no_grad():
+                x_hat = model(y).detach()
+
+            x_hat = torch.clamp(x_hat, 0.0, 1.0)
+
+            mask1, mask2 = N2N.generate_mask_pair(y)
+
+            y1 = N2N.generate_subimages(y, mask1)
```
Comment (Andrewwango) [inline deepinv/models/poisson2sparse.py:343]: nice integration with N2N, thanks

**Comment by jscanvic** [comment]
Ty for the review! Should be good
